### DISCUSSION TITLE
open points for our next techn. call (04/15/21)

### DISCUSSION THREAD

**Discussion by mgrossmann** [adesutherland/CREXX#27] [Q&A]
## open points for discussion
_(simply use the edit function, to add topics in this post)_

### architecture (MIG)
- unwanted dependencies between assembler/disassembler and interpreter; e.g. rxas.h needed by interpreter
 I would prefer a JRE/JDK like approach. (no dependencies at all) 
- `#ifndef NDEBUG` vs. `#ifdef DEBUG` **(AS note that NDEBUG is standard and works with ASSERT ...)**
-  build requirements -> REXX **(AS Noted! Might need to add a common build step (on Linux) before passing code to OSX / Windows build agents)**
-  how to pursue with code generation with REXX (here Regina)
       violation of C code rules?
       better solutions?
       **(AS Remember we are using RE2C and Lemon which are code generators too but lets discuss ...)**
- integration code generator is poc
-  argument handling;
currently every single argument is linked to an a style register, this makes it impossible to handle the arguments dynamically in the assembler code. e.g. loop over the count of arguments and do something useful with it. 
Is this meant to be handled by the compiler?  

**(AS Just saw your comment below re AMAP. I wrote my comment before reading. The point is that at runtime we can't map An to Rm if you see what I mean)** 

**(AS Register A0 has the number of arguments, and we need instructions like MOVE_IDX REG,REG and MAP_IDX REG,REG that moves/duplicated register via an index to the named register. e.g. MOVE_IDX R0,R1 moved the register with index stored in R1 to R0. The problem is that the human programmer / compiler needs to work out what the real register number is (i.e. A3 is really register R10, or whatever), so we will need to extend the Assembler to provide a special symbol .arg_index which converts to the index of A0. I.e. You need to do 
    LOAD R1,4   * I want A4 - note the compiler could probably work out the correct index
    ADD R1,R1,.arg_index  * Add the .arg_index ( = .globals + .locals) to R1
    MOVE_IDX R0,R1 * Get A4
    ... *Do stuff
    INC R1
    MOVE_IDX R0,R1 * Get A5  
)**

### instructions (MIG)
- `INC*`/`DEC*` for int values only? **(AS Yes)**
- `DEC0; IMASTER R0` vs. `calling master_int()`in `DEC0`
- `COPY_REG_REG` vs. `LOAD_REG_REG` **(AS Yes a duplicate!)**
- would like to have more information about `LINK` and `MAP` instructions **(AS TBC - these are the ways to access variables but terminology needs conforming)**
- regarding argument handling topic, I would suggest following new intstruchtions: **(AS See above)**
    `instr_f("amap", "Map op1 to arg register index in op2", OP_REG, OP_REG, OP_NONE);`
    `instr_f("amap", "Map op1 to arg register index  op2", OP_REG, OP_INT, OP_NONE);`

### instructions (PEJ)
- Implementation status
    89 functions are defined
    28 functions are not yet defined
  117 functions in total   

**(AS Boom! - we need to work out how to manage the list of them)**

- STRCAT functions not yet implemented due to missing "long string" capability.

- for String to integer conversions: do allow integers only in the string, or should we cut off the integer at the period char?
   example "3.14" could be either an error as no integer or is converted into a 3 **(AS I think that an error should be raised)**

- STR2INT_REG_nnn_nnn  String to Int op1 = op2[op3] and others having a string position parameter
  a/ is it an offset in the string (starting with zero) or
      a position (starting with 1) **(AS Very good question! Shall we be 1 based to be more REXX than C?)**
  b/ should a string length check occur? The compiler can't possibly do it, can it? **(AS we need an instruction STRLEN so that the compiler can insert check logic if it needs to)**

### error checking (PEJ)
- which level of error checking do we need in assembler?
   a/ DEBUG mode 
   b/ production mode
   c/ do we need to introduce more levels with a threshold command to enable/disable certain check levels?  

**(AS Perhaps DEBUG to do all checks, TRACE to Print Trace and Production?)**

### testing (PEJ)
  a/ instruction testing 
      we need voluntary testers of the assembler functions. The testing after implementation was very basic. More intensive testing is 
      necessary!
b/  maybe someone is willing to write a (Regina) Rexx to semi-automate the testing process

**(AS Agreed)**

- ...

@adesutherland @Peter-Jacob

**Comment by adesutherland** [comment]
Also would like to discuss the need to make the loader work as a link loader

**Reply by mgrossmann** [reply]
What is meant by this?

**Reply by adesutherland** [reply]
I mean it needs to be able to load multiple binary files (your .out files) and link then together so a procedure in one can call a procedure in another. That way we can have library of functions written in REXX.

Does that make sense?

**Reply by mgrossmann** [reply]
This makes perfectly sense to me.

**Reply by adesutherland** [reply]
I have added comments in line above

**Comment by mgrossmann** [comment]
Will create individual topics and close this thread. Thanks
